tests/returntypes/gradients/test_pulse_gradient_new.py

```diff
@@ -0,0 +1,900 @@
+        assert qml.math.isclose(r, jnp.cos(2 * p))
+        tapes, fn = stoch_pulse_grad(tape, num_samples=100, sampler_seed=7512)
+        assert len(tapes) == 200
+
+        res = fn(qml.execute(tapes, dev, None))
+        # The sampling of pwc functions does not automatically reduce to the analytically
+        # correct time integrals, leading to approximations -> coarse tolerance
+        assert qml.math.allclose(
+            res, -2 * jnp.sin(2 * p) * (T[1] - T[0]) / len(params[0]), atol=0.01
+        )
+
+    @pytest.mark.parametrize("t", [2.0, 3, (0.5, 0.6)])
+    def test_constant_commuting(self, t):
+        """Test that the derivative of a pulse generated by two constant commuting Hamiltonians
+        is computed correctly."""
+        import jax
+
+        jax.config.update("jax_enable_x64", True)
+        jnp = jax.numpy
+        params = [jnp.array(0.24), jnp.array(-0.672)]
+        T = t if isinstance(t, tuple) else (0, t)
+        op = qml.evolve(qml.pulse.constant * qml.PauliX(0) + qml.pulse.constant * qml.PauliY(1))(
+            params, t
+        )
+        tape = qml.tape.QuantumScript([op], [qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))])
+
+        dev = qml.device("default.qubit.jax", wires=2)
+        r = qml.execute([tape], dev, None)
+        # Effective rotation parameters
+        p = [_p * (T[1] - T[0]) for _p in params]
+        assert qml.math.isclose(r, jnp.cos(2 * p[0]) * jnp.cos(2 * p[1]))
+        tapes, fn = stoch_pulse_grad(tape)
+        assert len(tapes) == 4
+
+        res = fn(qml.execute(tapes, dev, None))
+        exp_grad = [
+            -2 * jnp.sin(2 * p[0]) * jnp.cos(2 * p[1]) * (T[1] - T[0]),
+            -2 * jnp.sin(2 * p[1]) * jnp.cos(2 * p[0]) * (T[1] - T[0]),
+        ]
+        assert qml.math.allclose(res, exp_grad)
+
+    def test_advanced_pulse(self):
+        """Test the derivative of a more complex pulse."""
+        import jax
+
+        jax.config.update("jax_enable_x64", True)
+        jnp = jax.numpy
+
+        ham = (
+            qml.pulse.constant * qml.PauliX(0)
+            + (lambda p, t: jnp.sin(p * t)) * qml.PauliZ(0)
+            + jnp.polyval * (qml.PauliY(0) @ qml.PauliY(1))
+        )
+        params = [jnp.array(1.51), jnp.array(-0.371), jnp.array([0.2, 0.2, -0.4])]
+        dev = qml.device("default.qubit.jax", wires=2)
+
+        @qml.qnode(dev, interface="jax")
+        def qnode(params):
+            qml.evolve(ham)(params, 0.4)
+            return qml.expval(qml.PauliY(0) @ qml.PauliX(1))
+
+        qnode.construct((params,), {})
+
+        num_samples = 100
+        tapes, fn = stoch_pulse_grad(
+            qnode.tape, argnums=[0, 1, 2], num_samples=num_samples, sampler_seed=7123
+        )
+        assert len(tapes) == 3 * 2 * num_samples
+
+        res = fn(qml.execute(tapes, dev, None))
+        exp_grad = jax.grad(qnode)(params)
+        assert all(qml.math.allclose(r, e, rtol=0.4) for r, e in zip(res, exp_grad))
+
+    def test_randomness(self):
```

Comment:
On second thought maybe actually better to remove them since what you are really testing here is that for those random seeds that you have chosen, the results differ to the ascribed tolerance, whereas it could be that for different random seeds, when the number of shots is large enough, the results _should_ be the same 🤔

Reply:
Mhm, yeah I mean this test is solid, because all randomness is fixed and there is no relying on shots. But I see your point. My idea was to make sure that the keys are not altered in a funky way outside of the sampling function `split_evol_ops` tested above, and to make sure that the full integration of fixing the randomness really does so. 
I don't have a strong opinion on this one though :)

Reply:
No super strong opinion on this one either, I think it won't fail in the future since everything is fixed nicely. 

It just bugs me conceptually that what we want is the function to give the same result independent of the random seed with enough samples, and what we are testing is that the function yields _different_ results for some specifically chosen random seeds in combination with a low enough sample size.

Reply:
Oh, interesting question! I think I wouldn't go there for this PR but I'll keep it in mind when I write the next similar test :)